app/validators/redirect_uri_validator.rb

```diff
@@ -27,14 +23,22 @@ def validate_each(record, attribute, value)
 
   private
 
-  def native_redirect_uri?(uri)
-    self.class.native_redirect_uri.present? && uri.to_s == self.class.native_redirect_uri.to_s
+  def oob_redirect_uri?(uri)
+    Doorkeeper::OAuth::NonStandard::IETF_WG_OAUTH2_OOB_METHODS.include?(uri)
   end
 
   def forbidden_uri?(uri)
     Doorkeeper.configuration.forbid_redirect_uri.call(uri)
   end
 
+  def relative_uri?(uri)
+    uri.scheme.nil? || (host_based_scheme?(uri) && uri.host.nil?)
+  end
+
+  def host_based_scheme?(uri)
+    %w[http https localhost].include?(uri.try(:scheme))
```

Comment:
The localhost here is because URI.parse('localhost:3000') interprets localhost as a scheme when presented without a scheme 🤷‍♂️ perhaps we want to handle this another way.

Reply:
I'm a bit concerned with this and I think we need to handle differently.

From what I see, URI.parse creates different objects for known schemes:

```
[9] pry(main)> URI.parse('localhost:3000')
=> #<URI::Generic localhost:3000>
[10] pry(main)> URI.parse('http://localhost:3000')
=> #<URI::HTTP http://localhost:3000>
[11] pry(main)> URI.parse('https://localhost:3000')
=> #<URI::HTTPS https://localhost:3000>
[12] pry(main)> URI.parse('myapp://localhost:3000')
=> #<URI::Generic myapp://localhost:3000>
[13] pry(main)> URI.parse('myapp:/localhost:3000')
=> #<URI::Generic myapp:/localhost:3000>
[14] pry(main)> URI.parse('ftp://localhost:3000')
=> #<URI::FTP ftp://localhost:3000/>
```

I'm not sure it will help, but maybe splitting the validation between HTTP, HTTPS, Generic (and maybe not allowing other types of URIs like FTP) would simplify a bit

Reply:
I have been thinking about this a bit more. I was tempted to just remove the localhost and change the test. The way the URI library works seems consistent with other URI libraries in other languages and perhaps it's just a misconfiguration by the developer we don't actually need to conscientiously prevent. 

I can switch to checking the type returned by URI.parse but do we consider dynamic type checking (instance_of? etc) a code smell in ruby? (in other languages it would be considered so).

After those thoughts I started to actually wonder if we need the relative_uri check at all but it does indeed state [here in rfc6749](https://tools.ietf.org/html/rfc6749#section-3.1.2) that that is the case. However, it can only conceivably apply in the case of http and https URI.

I suggest that scheme is always explicitly stated, as my interpretation is that that must always be that case anyway.

so maybe

```

def relative_uri?(uri)
  scheme_requires_authority?(uri) && uri.host.nil?
end 

def scheme_requires_authority?(uri)
  ['http', 'https'].include?(uri.scheme)
end

```

Do we want to explicitly forbid uri without schemes? [here in rfc6749 section 3.1.2](https://tools.ietf.org/html/rfc6749#section-3.1.2) in conjunction with [rfc3886 section 4.3](https://tools.ietf.org/html/rfc3986#section-4.3) implies this in my view.

``` 
record.errors.add(attribute, :fragment_present) unless uri.fragment.nil?
record.errors.add(attribute, :scheme_missing) if uri.scheme.nil?
```

Reply:
Actually perhaps all this goes away if we just do:

```
record.errors.add(attribute, :relative_uri) if uri.scheme.nil? && uri.host.nil?
``

Reply:
I'm not very confident on the latest approach. This would let pass things like:
```ruby
uri = URI.parse('app.com:80')
uri.scheme.nil? && uri.host.nil? # => false, scheme is there
```
I noticed though, that when such URI is passed, the value of `opaque` is set (instead of port):
```ruby
URI.parse('app.com:80').opaque
# => "80"
```

So I'm thinking about extending what you suggested:
```ruby
def validate(uri)
  uri = URI.parse(uri)
  !uri.scheme.nil? && !uri.host.nil? && uri.fragment.nil? && uri.opaque.nil?
end

[28] pry(main)> validate('app.com:80')
=> false
[29] pry(main)> validate('app.com:/80')
=> false
[30] pry(main)> validate('app.com://80')
=> true
[31] pry(main)> validate('https://80')
=> true
[32] pry(main)> validate('http://80')
=> true
[33] pry(main)> validate('http://80?test=1')
=> true
[34] pry(main)> validate('http://80?test=1#oops')
=> false
[35] pry(main)> validate('app.com://80?test=1#oops')
=> false
[36] pry(main)> validate('app.com://80?test=1')
=> true
[37] pry(main)> validate('locahost:9000')
=> false
```

The only caveat is that this rejects `app.com:/80`, which I think it makes sense. WDYT?

Reply:
The most immediate problem in my view with this validation is that the _most_ rfc correct presentation for apps with native scheme will also be rejected. 

```
validate('com.domain.app:/oauth2')
=> false
```

`com.domain.app://oauth2` is not technically correct; it's a scheme and host not a scheme and path but we have decided to allow it as some native apps do this incorrectly but we cannot expect app developers to use this incorrect presentation out of necessity.

With the rules of rfc alone it is difficult to resolve an interpretation in every case or such a simple presentation. It simply isn't enough information. So as humans we  might use the appearance as a hint as to what it might actually mean. We see the number and interpret it as a port, we see localhost and assume a host over a scheme. 

We are not the first with this [problem](https://bitbucket.org/openid/connect/issues/625/discovery-223-examplecom-8080-has-a-scheme)

I think since we are only validating if it is relative here I think it is ok, do we really want to validate here that somebody has incorrectly presented a host based URI without authority marker '//' here?
 
We could attempt to pick-up this misconfiguration up earlier. Warn if redirect_uri look like host based uri without schemes.

I will continue to think on this.

Reply:
Thanks for digging deeper on this 👏 

I have been misinterpreting this all along 🤦‍♂️! In my mind such URI `com.domain.app:/oauth2` didn't make sense because I was used to thinking that URIs should look like `scheme://something` - with `://` always there.

But indeed, like you said, my interpretation is `scheme` and `host` and not `scheme` and `path`.

> I think since we are only validating if it is relative here I think it is ok, do we really want to validate here that somebody has incorrectly presented a host based URI without authority marker '//' here?

If I understand correctly your question, after your explanation I think that no, we should not require the presence/absence of the authority marker for **generic URIs**. In short, `com.domain.app://oauth2` and `com.domain.app:/oauth2` are both valid.

Do you agree that we should at least require a `path` to avoid confusion? In other words, a path-rootless URI like`com.domain.app:oauth2` should be invalid (due to potential confusion with `domain:port`)?

Note: I hope we also agree that we are talking only about generic URIs and we should treat differently in the code the validation of http/https/generic schemes, for maintainability purposes.

Reply:
It might be that looking for the existence of a path works for most cases but I don't see anything that explicitly excludes a pathless URI for use as a redirect_url. It seems to me to be a plausible use case. Especially in the case of _Loopback Interface Redirection_ as discussed in [rfc8252#section-7.3](https://tools.ietf.org/html/rfc8252#section-7.3) where a redirect uri with a form like `http://127.0.0.1:9876/` seems highly plausible since the port will only be opened briefly on the client system in order to receive the callback and the closed again, it simply doesn't need a path.

I think what I am saying is that it makes it very hard to protect integrators from the myriad of misconfigurations possible in that validation without also closing off perfectly valid use cases in this validation. 

doorkeeper has historically allowed simply `localhost:80` and similar but I think we it should be discouraged, it should be presented as `http://localhost:80`.

I am coming around to the view that we should, on loading the configuration heuristically map common misconfigurations: 

`localhost:8080` => `http://localhost:8080` 

_and_ print a visible warning explaining that this happened and implementors should use the fully qualified URI to avoid problems and undefined behaviour. We can also update the document to make it clear integrators need to be explicit. 

In this way we can try to support existing deployed misconfigurations if that is deemed important.

Reply:
Thanks for your reply.

> doorkeeper has historically allowed simply localhost:80 and similar but I think we it should be discouraged, it should be presented as http://localhost:80.

I am unable to create an application with simply `localhost:80`, on 5.1 I get `Redirect URI must be an absolute URI.`. It requires `https` too by default, and I know that this can be changed via setting. We should keep that.

> In this way we can try to support existing deployed misconfigurations if that is deemed important.

Do you have an example of such misconfiguration? I might be missing something here.

Additionally, I'm also able to register redirect URI as `com.example.app://test` but not `com.example.app:/test`. Now that I have full understanding of what the latter means, I do believe we should allow such validation, but that should not come at the cost of allowing `localhost:80` for two reasons:
1. backwards compatibility, in other words no surprises when updating doorkeeper.
2. it creates questions when redirecting, like which scheme should it use.

I'd like to keep that constraint, if you agree

Reply:
> Do you have an example of such misconfiguration? I might be missing something here. 

I just meant that I consider the simple `localhost:80` etc a misconfiguration by strict interpretation of the rfcs but I understand that the reality is that there are likely many instances in use like this already. 

OK. I'll have another run at this part. I'm currently on holiday but I am back on the 29 April 2019 so it will be around then.